libgadget/treewalk_kernel.cu

```diff
@@ -0,0 +1,549 @@
+{
+    const ForceTree * tree = lv->tw->tree;
+    const double BoxSize = tree->BoxSize;
+
+    /*Tree-opening constants*/
+    const double cellsize = GRAV_GET_PRIV(lv->tw)->cellsize;
+    const double rcut = GRAV_GET_PRIV(lv->tw)->Rcut;
+    const double rcut2 = rcut * rcut;
+    const double aold = TreeParams_ptr->ErrTolForceAcc * input->OldAcc;
+    const int TreeUseBH = TreeParams_ptr->TreeUseBH;
+    double BHOpeningAngle2 = TreeParams_ptr->BHOpeningAngle * TreeParams_ptr->BHOpeningAngle;
+    /* Enforce a maximum opening angle even for relative acceleration criterion, to avoid
+     * pathological cases. Default value is 0.9, from Volker Springel.*/
+    if(TreeUseBH == 0)
+        BHOpeningAngle2 = TreeParams_ptr->MaxBHOpeningAngle * TreeParams_ptr->MaxBHOpeningAngle;
+
+    /*Input particle data*/
+    const double * inpos = input->base.Pos;
+
+    TreeWalkResultChildren output_children[1] = {{0}};
+
+    /*Start the tree walk*/
+    int listindex, ninteractions=0;
+
+    /* Primary treewalk only ever has one nodelist entry*/
+    for(listindex = 0; listindex < NODELISTLENGTH; listindex++)
+    {
+        int numcand = 0;
+        /* Use the next node in the node list if we are doing a secondary walk.
+         * For a primary walk the node list only ever contains one node. */
+        int no = input->base.NodeList[listindex];
+        int startno = no;
+        if(no < 0)
+            break;
+
+        while(no >= 0)
+        {
+            /* The tree always walks internal nodes*/
+            struct NODE *nop = &tree->Nodes[no];
+
+            if(lv->mode == TREEWALK_GHOSTS && nop->f.TopLevel && no != startno)  /* we reached a top-level node again, which means that we are done with the branch */
+                break;
+
+            int i;
+            double dx[3];
+            for(i = 0; i < 3; i++)
+                dx[i] = NEAREST(nop->mom.cofm[i] - inpos[i], BoxSize);
+            const double r2 = dx[0] * dx[0] + dx[1] * dx[1] + dx[2] * dx[2];
+
+            /* Discard this node, move to sibling*/
+            if(shall_we_discard_node_device(nop->len, r2, nop->center, inpos, BoxSize, rcut, rcut2))
+            {
+                no = nop->sibling;
+                /* Don't add this node*/
+                continue;
+            }
+
+            /* This node accelerates the particle directly, and is not opened.*/
+            int open_node = shall_we_open_node_device(nop->len, nop->mom.mass, r2, nop->center, inpos, BoxSize, aold, TreeUseBH, BHOpeningAngle2);
+            
+            if(!open_node)
+            {
+                /* ok, node can be used */
+                no = nop->sibling;
+                if(lv->mode != TREEWALK_TOPTREE) {
+                    /* Compute the acceleration and apply it to the output structure*/
+                    apply_accn_to_output_device(output, dx, r2, nop->mom.mass, cellsize);
```

Comment:
Should be                     apply_accn_to_output_device(output, dx, r2, nop->mom.mass, cellsize, 0);